pkg/arvo/lib/pull-hook.hoon

```diff
@@ -295,72 +312,310 @@
+      [cards this]
+    ::
+    ++  on-leave
+      |=  =path
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-leave:og path)
+      [cards this]
+    ::
+    ++  on-arvo
+      |=  [=wire =sign-arvo]
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-arvo:og wire sign-arvo)
+      [cards this]
+    ::
+    ++  on-fail
+      |=  [=term =tang]
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-fail:og term tang)
+      [cards this]
+    ::
+    ++  on-peek   
+      |=  =path
+      ^-  (unit (unit cage))
+      ?:  =(/x/dbug/state path)
+        ``noun+(slop !>(state(inner-state *vase)) on-save:og)
+      ?.  =(/x/tracking path)
+        (on-peek:og path)
+      ``noun+!>(~(key by tracking))
     --
   |_  =bowl:gall
   +*  og   ~(. pull-hook bowl)
+      io   ~(. agentio bowl)
+      pass  pass:io
+      virt  ~(. pull-hook-virt bowl)
+      ver  ~(. versioning [bowl [update-mark version min-version]:config])
+  ::
+  ++  track-engine
+    |_  [cards=(list card) rid=resource =ship =status gone=_|]
+    ::  +|  %init: state machine setup and manipulation
+    ::
+    ++  tr-core  .
+    ++  tr-abed
+      |=  r=resource
+      =/  [s=^ship sta=^status]
+        (~(got by tracking) r)
+      tr-core(rid r, ship s, status sta)
+    ::
+    ++  tr-abet
+      ^-  [[(list card) _pull-hook] _state]
+      =.  tracking
+        ?:  gone
+          (~(del by tracking) rid)
+        (~(put by tracking) rid [ship status])
+      [[(flop cards) pull-hook] state]
+    ::
+    ++  tr-emit
+      |=  =card
+      tr-core(cards [card cards])
+    ::
+    ++  tr-emis
+      |=  caz=(list card)
+      tr-core(cards (welp (flop cards) cards))
+    ::
+    ++  tr-ap-og
+      |=  ap=_^?(|.(*(quip card _pull-hook)))
+      =^  caz  pull-hook
+        (ap)
+      (tr-emis caz)
+    ::  +|  %sign: sign handling
+    ::
+    ::
+    ++  tr-sign
+      |=  [=sign:agent:gall versioned=?]
+      ?+   -.sign  !!
+        %kick       tr-kick
+        %watch-ack  (tr-wack +.sign versioned)
+        %fact       (tr-fact +.sign)
+      ==
+    ::
+    ::
+    ++  tr-wack
+      |=  [tan=(unit tang) versioned=?]
+      ?~  tan  tr-core
+      ?.  versioned
+        (tr-ap-og:tr-cleanup |.((on-pull-nack:og rid u.tan)))
+      =/  pax
+        (kick-mule:virt rid |.((on-pull-kick:og rid)))
+      ?~  pax  tr-failed-kick
+      ?~  u.pax  tr-cleanup
+      (tr-watch-unver u.u.pax)
+    ::
+    ++  tr-kick
+      ?.  ?=(%active -.status)  tr-core
+      =/  pax
+        (kick-mule:virt rid |.((on-pull-kick:og rid)))
+      ?~  pax  tr-failed-kick
+      ?~  u.pax  tr-cleanup
+      (tr-watch u.u.pax)
+    ::
+    ++  tr-fact
+      |=  =cage
+      ?:  ?=(%version p.cage)
+        =/  req-ver=@ud
+          !<(@ud q.cage)
+        ?:  (lth req-ver min-version.config)
+          (tr-suspend-pub-ver min-version.config)
+        (tr-suspend-sub-ver req-ver)
```

Comment:
Why do we suspend on sub-ver here? Doesn't that mean we suspend in all cases upon receiving a version fact?

Reply:
Yes, because a `version` fact is only sent if the publisher cannot convert it's facts to the requested version. It's always followed by a kick.